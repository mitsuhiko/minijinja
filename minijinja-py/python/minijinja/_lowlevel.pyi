from typing import Any
from . import Environment
from typing_extensions import final

@final
class State:
    name: str
    env: Environment
    current_block: str | None
    auto_escape: bool

    def lookup(self, name: str) -> Any: ...
