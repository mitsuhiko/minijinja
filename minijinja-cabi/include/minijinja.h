/*
 * MiniJinja for C
 *
 * Copyright 2024 Armin Ronacher. MIT License.
 */

#ifndef _minijinja_h_included
#define _minijinja_h_included

#pragma once

#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>

#ifndef MINIJINJA_API
#  define MINIJINJA_API
#endif


/*
 The kind of error that occurred.
 */
typedef enum mj_err_kind {
  MJ_ERR_KIND_NON_PRIMITIVE,
  MJ_ERR_KIND_NON_KEY,
  MJ_ERR_KIND_INVALID_OPERATION,
  MJ_ERR_KIND_SYNTAX_ERROR,
  MJ_ERR_KIND_TEMPLATE_NOT_FOUND,
  MJ_ERR_KIND_TOO_MANY_ARGUMENTS,
  MJ_ERR_KIND_MISSING_ARGUMENT,
  MJ_ERR_KIND_UNKNOWN_FILTER,
  MJ_ERR_KIND_UNKNOWN_FUNCTION,
  MJ_ERR_KIND_UNKNOWN_TEST,
  MJ_ERR_KIND_UNKNOWN_METHOD,
  MJ_ERR_KIND_BAD_ESCAPE,
  MJ_ERR_KIND_UNDEFINED_ERROR,
  MJ_ERROR_KIND_BAD_SERIALIZTION,
  MJ_ERR_KIND_BAD_INCLUDE,
  MJ_ERR_KIND_EVAL_BLOCK,
  MJ_ERR_KIND_CANNOT_UNPACK,
  MJ_ERR_KIND_WRITE_FAILURE,
  MJ_ERR_KIND_UNKNOWN,
} mj_err_kind;

/*
 Controls the undefined behavior of the engine.
 */
typedef enum mj_undefined_behavior {
  /*
   The default, somewhat lenient undefined behavior.
   */
  MJ_UNDEFINED_BEHAVIOR_LENIENT,
  /*
   Complains very quickly about undefined values.
   */
  MJ_UNDEFINED_BEHAVIOR_STRICT,
  /*
   Like Lenient, but also allows chaining of undefined lookups.
   */
  MJ_UNDEFINED_BEHAVIOR_CHAINABLE,
} mj_undefined_behavior;

/*
 The kind of a value.
 */
typedef enum mj_value_kind {
  MJ_VALUE_KIND_UNDEFINED,
  MJ_VALUE_KIND_NONE,
  MJ_VALUE_KIND_BOOL,
  MJ_VALUE_KIND_NUMBER,
  MJ_VALUE_KIND_STRING,
  MJ_VALUE_KIND_BYTES,
  MJ_VALUE_KIND_SEQ,
  MJ_VALUE_KIND_MAP,
  MJ_VALUE_KIND_ITERABLE,
  MJ_VALUE_KIND_PLAIN,
  MJ_VALUE_KIND_INVALID,
} mj_value_kind;

/*
 Pointer to a MiniJinja environment.
 */
typedef struct mj_env mj_env;

/*
 Helps iterating over a value.
 */
typedef struct mj_value_iter mj_value_iter;

/*
 Opaque value type.
 */
typedef struct mj_value {
  uintptr_t _opaque[3];
} mj_value;

/*
 Allows one to override the syntax elements.
 */
typedef struct mj_syntax_config {
  const char *block_start;
  const char *block_end;
  const char *variable_start;
  const char *variable_end;
  const char *comment_start;
  const char *comment_end;
  const char *line_statement_prefix;
  const char *line_comment_prefix;
} mj_syntax_config;

#ifdef __cplusplus
extern "C" {
#endif // __cplusplus

/*
 Registers a template with the environment.
 */
MINIJINJA_API bool mj_env_add_template(struct mj_env *env, const char *name, const char *source);

/*
 Clears all templates.
 */
MINIJINJA_API bool mj_env_clear_templates(struct mj_env *env);

/*
 Evaluate an expression.
 */
MINIJINJA_API
struct mj_value mj_env_eval_expr(const struct mj_env *env,
                                 const char *expr,
                                 struct mj_value ctx);

/*
 Frees a MiniJinja environment.
 */
MINIJINJA_API void mj_env_free(struct mj_env *env);

/*
 Allocates a new and empty MiniJinja environment.
 */
MINIJINJA_API struct mj_env *mj_env_new(void);

/*
 Removes a template from the environment.
 */
MINIJINJA_API bool mj_env_remove_template(struct mj_env *env, const char *name);

/*
 Renders a template from a named string.
 */
MINIJINJA_API
char *mj_env_render_named_str(const struct mj_env *env,
                              const char *name,
                              const char *source,
                              struct mj_value ctx);

/*
 Renders a template registered on the environment.
 */
MINIJINJA_API
char *mj_env_render_template(const struct mj_env *env,
                             const char *name,
                             struct mj_value ctx);

/*
 Enables or disables debug mode.
 */
MINIJINJA_API void mj_env_set_debug(struct mj_env *env, bool val);

/*
 Preserve the trailing newline when rendering templates.
 */
MINIJINJA_API void mj_env_set_keep_trailing_newline(struct mj_env *env, bool val);

/*
 Enables or disables the `lstrip_blocks` feature.
 */
MINIJINJA_API void mj_env_set_lstrip_blocks(struct mj_env *env, bool val);

/*
 Changes the recursion limit.
 */
MINIJINJA_API void mj_env_set_recursion_limit(struct mj_env *env, uint32_t val);

/*
 Reconfigures the syntax.
 */
MINIJINJA_API
bool mj_env_set_syntax_config(struct mj_env *env,
                              const struct mj_syntax_config *syntax);

/*
 Enables or disables the `trim_blocks` feature.
 */
MINIJINJA_API void mj_env_set_trim_blocks(struct mj_env *env, bool val);

/*
 Reconfigures the undefined behavior.
 */
MINIJINJA_API
void mj_env_set_undefined_behavior(struct mj_env *env,
                                   enum mj_undefined_behavior val);

/*
 Clears the current error.
 */
MINIJINJA_API void mj_err_clear(void);

/*
 Returns the error's description if there is an error.
 */
MINIJINJA_API const char *mj_err_get_detail(void);

/*
 Returns the error's kind
 */
MINIJINJA_API enum mj_err_kind mj_err_get_kind(void);

/*
 Returns the error's current line.
 */
MINIJINJA_API uint32_t mj_err_get_line(void);

/*
 Returns the error's current template.
 */
MINIJINJA_API const char *mj_err_get_template_name(void);

/*
 Returns `true` if there is currently an error.
 */
MINIJINJA_API bool mj_err_is_set(void);

/*
 Prints the error to stderr.
 */
MINIJINJA_API bool mj_err_print(void);

/*
 Frees an engine allocated string.
 */
MINIJINJA_API void mj_str_free(char *s);

/*
 Sets the syntax to defaults.
 */
MINIJINJA_API void mj_syntax_config_default(struct mj_syntax_config *syntax);

/*
 Appends a value to a list
 */
MINIJINJA_API bool mj_value_append(struct mj_value *slf, struct mj_value value);

/*
 Extracts a float from the value
 */
MINIJINJA_API double mj_value_as_f64(struct mj_value value);

/*
 Extracts an integer from the value
 */
MINIJINJA_API int64_t mj_value_as_i64(struct mj_value value);

/*
 Extracts an unsigned integer from the value
 */
MINIJINJA_API uint64_t mj_value_as_u64(struct mj_value value);

/*
 Debug prints a value to stderr
 */
MINIJINJA_API void mj_value_dbg(struct mj_value value);

/*
 Decrements the refcount
 */
MINIJINJA_API void mj_value_decref(struct mj_value *value);

/*
 Looks up an element by an integer index in a list of object
 */
MINIJINJA_API struct mj_value mj_value_get_by_index(struct mj_value value, uint64_t idx);

/*
 Looks up an element by a string index in an object.
 */
MINIJINJA_API struct mj_value mj_value_get_by_str(struct mj_value value, const char *key);

/*
 Looks up an element by a vaue
 */
MINIJINJA_API struct mj_value mj_value_get_by_value(struct mj_value value, struct mj_value key);

/*
 Returns the value kind.
 */
MINIJINJA_API enum mj_value_kind mj_value_get_kind(struct mj_value value);

/*
 Increments the refcount
 */
MINIJINJA_API void mj_value_incref(struct mj_value *value);

/*
 Checks if the value is numeric
 */
MINIJINJA_API bool mj_value_is_number(struct mj_value value);

/*
 Checks if the value is truthy
 */
MINIJINJA_API bool mj_value_is_true(struct mj_value value);

/*
 Ends the iteration and deallocates the iterator
 */
MINIJINJA_API void mj_value_iter_free(struct mj_value_iter *iter);

/*
 Yields the next value from the iterator.
 */
MINIJINJA_API bool mj_value_iter_next(struct mj_value_iter *iter, struct mj_value *val_out);

/*
 Returns the length of the object
 */
MINIJINJA_API uint64_t mj_value_len(struct mj_value value);

/*
 Creates a new boolean value
 */
MINIJINJA_API struct mj_value mj_value_new_bool(bool value);

/*
 Creates a new f32 value
 */
MINIJINJA_API struct mj_value mj_value_new_f32(float value);

/*
 Creates a new f64 value
 */
MINIJINJA_API struct mj_value mj_value_new_f64(double value);

/*
 Creates a new i32 value
 */
MINIJINJA_API struct mj_value mj_value_new_i32(int32_t value);

/*
 Creates a new i64 value
 */
MINIJINJA_API struct mj_value mj_value_new_i64(int64_t value);

/*
 Creates an empty list value
 */
MINIJINJA_API struct mj_value mj_value_new_list(void);

/*
 Creates a new none value.
 */
MINIJINJA_API struct mj_value mj_value_new_none(void);

/*
 Creates an empty object value
 */
MINIJINJA_API struct mj_value mj_value_new_object(void);

/*
 Creates a new string value
 */
MINIJINJA_API struct mj_value mj_value_new_string(const char *s);

/*
 Creates a new u32 value
 */
MINIJINJA_API struct mj_value mj_value_new_u32(uint32_t value);

/*
 Creates a new u64 value
 */
MINIJINJA_API struct mj_value mj_value_new_u64(uint64_t value);

/*
 Creates a new undefined value.
 */
MINIJINJA_API struct mj_value mj_value_new_undefined(void);

/*
 Inserts a key into an object value.
 */
MINIJINJA_API
bool mj_value_set_key(struct mj_value *slf,
                      struct mj_value key,
                      struct mj_value value);

/*
 Inserts a string key into an object value.
 */
MINIJINJA_API
bool mj_value_set_string_key(struct mj_value *slf,
                             const char *key,
                             struct mj_value value);

/*
 Converts the value into a string.
 */
MINIJINJA_API char *mj_value_to_str(struct mj_value value);

/*
 Iterates over the value.
 */
MINIJINJA_API struct mj_value_iter *mj_value_try_iter(struct mj_value value);

#ifdef __cplusplus
}  // extern "C"
#endif  // __cplusplus

#endif  /* _minijinja_h_included */
