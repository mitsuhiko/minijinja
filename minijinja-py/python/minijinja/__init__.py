from . import _lowlevel


__all__ = ["Environment", "safe", "escape", "render_str", "eval_expr"]


class Environment(_lowlevel.Environment):
    """Represents a MiniJinja environment"""
    def __new__(cls, *args, **kwargs):
        # `_lowlevel.Environment` does not accept any arguments
        return super().__new__(cls)

    def __init__(
        self,
        loader=None,
        templates=None,
        filters=None,
        tests=None,
        globals=None,
        debug=True,
        fuel=None,
        auto_escape_callback=None,
    ):
        super()
        if loader is not None:
            if templates:
                raise TypeError("Cannot set loader and templates at the same time")
            self.loader = loader
        elif templates is not None:
            self.loader = dict(templates).get
        if fuel is not None:
            self.fuel = fuel
        if filters:
            for name, callback in filters.items():
                self.add_filter(name, callback)
        if tests:
            for name, callback in tests.items():
                self.add_test(name, callback)
        if globals is not None:
            for name, value in globals.items():
                self.add_global(name, value)
        self.debug = debug
        if auto_escape_callback is not None:
            self.auto_escape_callback = auto_escape_callback


def render_str(__source, __name=None, **context):
    """Shortcut to render a string with the default environment."""
    return Environment().render_str(__source, __name, **context)


def eval_expr(__expr, **context):
    """Evaluate an expression with the default environment."""
    return Environment().eval_expr(__expr, **context)


try:
    from markupsafe import escape, Markup
except ImportError:
    from html import escape as _escape

    class Markup(str):
        def __html__(self):
            return self

    def escape(value):
        callback = getattr(value, "__html__", None)
        if callback is not None:
            return callback()
        return Markup(_escape(str(value)))


def safe(s):
    """Marks a string as safe."""
    return Markup(s)
