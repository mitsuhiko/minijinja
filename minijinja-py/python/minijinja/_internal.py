# This file contains functions that the rust module imports.
from . import TemplateError, safe


def make_error(info):
    # Internal utility function used by the rust binding to create a template error
    # with info object.  We cannot directly create an error on the Rust side because
    # we want to subclass the runtime error, but on the limited abi it's not possible
    # to create subclasses (yet!)
    err = TemplateError(info.description)
    err._info = info
    return err


# used by the rust runtime to mark something as safe
mark_safe = safe
