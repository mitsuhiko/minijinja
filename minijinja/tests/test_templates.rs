#![cfg(all(
    feature = "multi_template",
    feature = "macros",
    feature = "builtins",
    feature = "adjacent_loop_items"
))]
use std::collections::BTreeMap;
use std::fmt::Write;
use std::sync::Arc;
use std::{env, fs};

use insta::assert_snapshot;
use minijinja::value::{Enumerator, Object, ObjectRepr, Value};
use minijinja::{context, render, Environment, Error, ErrorKind, State};

use similar_asserts::assert_eq;

#[test]
fn test_vm() {
    let mut refs = Vec::new();
    insta::glob!("inputs/refs/*", |entry| {
        let filename = entry.file_name().unwrap();
        let filename = filename.to_str().unwrap();
        if filename.ends_with(".txt") || filename.ends_with(".html") {
            let source = fs::read_to_string(entry).unwrap();
            refs.push((entry.to_path_buf(), source));
        }
    });

    insta::glob!("inputs/*", |path| {
        if !path.metadata().unwrap().is_file() {
            return;
        }
        let filename = path.file_name().unwrap().to_str().unwrap();
        let contents = std::fs::read_to_string(path).unwrap();
        let mut iter = contents.splitn(2, "\n---\n");
        let mut env = Environment::new();
        let ctx: serde_json::Value = serde_json::from_str(iter.next().unwrap()).unwrap();

        for (path, source) in &refs {
            let ref_filename = path.file_name().unwrap().to_str().unwrap();
            env.add_template(ref_filename, source).unwrap();
        }

        let content = iter.next().unwrap();
        let rendered = if let Err(err) = env.add_template(filename, content) {
            let mut rendered = format!("!!!SYNTAX ERROR!!!\n\n{err:#?}\n\n");
            writeln!(rendered, "{err:#}").unwrap();
            rendered
        } else {
            let template = env.get_template(filename).unwrap();

            match template.render(&ctx) {
                Ok(mut rendered) => {
                    rendered.push('\n');
                    rendered
                }
                Err(err) => {
                    let mut rendered = format!("!!!ERROR!!!\n\n{err:#?}\n\n");

                    writeln!(rendered, "{err:#}").unwrap();
                    let mut err = &err as &dyn std::error::Error;
                    while let Some(next_err) = err.source() {
                        writeln!(rendered).unwrap();
                        writeln!(rendered, "caused by: {next_err:#}").unwrap();
                        err = next_err;
                    }

                    rendered
                }
            }
        };

        insta::with_settings!({
            info => &ctx,
            description => content.trim_end(),
            omit_expression => true
        }, {
            insta::assert_snapshot!(&rendered);
        });
    });
}

#[test]
fn test_vm_block_fragments() {
    let mut refs = Vec::new();
    insta::glob!("fragment-inputs/refs/*", |entry| {
        let filename = entry.file_name().unwrap();
        let filename = filename.to_str().unwrap();
        if filename.ends_with(".txt") || filename.ends_with(".html") {
            let source = fs::read_to_string(entry).unwrap();
            refs.push((entry.to_path_buf(), source));
        }
    });

    insta::glob!("fragment-inputs/*", |path| {
        if !path.metadata().unwrap().is_file() {
            return;
        }
        let filename = path.file_name().unwrap().to_str().unwrap();
        let contents = std::fs::read_to_string(path).unwrap();
        let mut iter = contents.splitn(2, "\n---\n");
        let mut env = Environment::new();
        let ctx: serde_json::Value = serde_json::from_str(iter.next().unwrap()).unwrap();

        for (path, source) in &refs {
            let ref_filename = path.file_name().unwrap().to_str().unwrap();
            env.add_template(ref_filename, source).unwrap();
        }

        let content = iter.next().unwrap();
        let rendered = if let Err(err) = env.add_template(filename, content) {
            let mut rendered = format!("!!!SYNTAX ERROR!!!\n\n{err:#?}\n\n");
            writeln!(rendered, "{err:#}").unwrap();
            rendered
        } else {
            let template = env.get_template(filename).unwrap();

            match template
                .eval_to_state(&ctx)
                .and_then(|mut x| x.render_block("fragment"))
            {
                Ok(mut rendered) => {
                    rendered.push('\n');
                    rendered
                }
                Err(err) => {
                    let mut rendered = format!("!!!ERROR!!!\n\n{err:#?}\n\n");

                    writeln!(rendered, "{err:#}").unwrap();
                    let mut err = &err as &dyn std::error::Error;
                    while let Some(next_err) = err.source() {
                        writeln!(rendered).unwrap();
                        writeln!(rendered, "caused by: {next_err:#}").unwrap();
                        err = next_err;
                    }

                    rendered
                }
            }
        };

        insta::with_settings!({
            info => &ctx,
            description => content.trim_end(),
            omit_expression => true
        }, {
            insta::assert_snapshot!(&rendered);
        });
    });
}

#[test]
fn test_custom_filter() {
    fn test_filter(_: &State, value: String) -> Result<String, Error> {
        Ok(format!("[{value}]"))
    }

    let mut ctx = BTreeMap::new();
    ctx.insert("var", 42);

    let mut env = Environment::new();
    env.add_filter("test", test_filter);
    env.add_template("test", "{{ var|test }}").unwrap();
    let tmpl = env.get_template("test").unwrap();
    let rv = tmpl.render(&ctx).unwrap();
    assert_eq!(rv, "[42]");
}

#[test]
fn test_items_and_dictsort_with_structs() {
    #[derive(Debug, Clone)]
    struct MyStruct;

    impl Object for MyStruct {
        fn get_value(self: &Arc<Self>, name: &Value) -> Option<Value> {
            match name.as_str()? {
                "a" => Some(Value::from("A")),
                "b" => Some(Value::from("B")),
                _ => None,
            }
        }

        fn enumerate(self: &Arc<Self>) -> Enumerator {
            Enumerator::Str(&["b", "a"])
        }
    }

    insta::assert_snapshot!(
        minijinja::render!("{{ x|items }}", x => Value::from_object(MyStruct)),
        @r###"[["b", "B"], ["a", "A"]]"###
    );
    insta::assert_snapshot!(
        minijinja::render!("{{ x|dictsort }}", x => Value::from_object(MyStruct)),
        @r###"[["a", "A"], ["b", "B"]]"###
    );
}

#[test]
fn test_urlencode_with_struct() {
    #[derive(Debug, Clone)]
    struct MyStruct;

    impl Object for MyStruct {
        fn get_value(self: &Arc<Self>, name: &Value) -> Option<Value> {
            match name.as_str() {
                Some("a") => Some(Value::from("a 1")),
                Some("b") => Some(Value::from("b 2")),
                _ => None,
            }
        }

        fn enumerate(self: &Arc<Self>) -> Enumerator {
            Enumerator::Str(&["a", "b"])
        }
    }

    insta::assert_snapshot!(
        minijinja::render!("{{ x|urlencode }}", x => Value::from_object(MyStruct)),
        @"a=a%201&b=b%202"
    );
}

#[test]
fn test_single() {
    let mut env = Environment::new();
    env.add_template("simple", "Hello {{ name }}!").unwrap();
    let tmpl = env.get_template("simple").unwrap();
    let rv = tmpl.render(context!(name => "Peter")).unwrap();
    assert_eq!(rv, "Hello Peter!");
}

#[test]
fn test_values_scientific_notation() {
    let mut env = Environment::new();
    env.add_template("sci1", "VALUE = {{ value or -12.4E-4 }}")
        .unwrap();
    let tmpl = env.get_template("sci1").unwrap();
    let rv = tmpl.render(context!(value => -12.4E-3)).unwrap();
    assert_eq!(rv, "VALUE = -0.0124");
    let rv = tmpl.render(context!());
    // assert_eq!(rv, "VALUE = -0.00124");
    assert!(rv.is_ok());

    env.add_template("sci2", "VALUE = {{ value or 1.4E4 }}")
        .unwrap();
    let tmpl = env.get_template("sci2").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());

    env.add_template("sci3", "VALUE = {{ value or 1.4e+4}}")
        .unwrap();
    let tmpl = env.get_template("sci3").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());

    env.add_template("sci4", "VALUE = {{ 1.4+4}}").unwrap();
    let tmpl = env.get_template("sci4").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());

    env.add_template("sci5", "VALUE = {{ 1.4+1E-1}}").unwrap();
    let tmpl = env.get_template("sci5").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());

    env.add_template("sci6", "VALUE = {{ 1.0E0+1.0}}").unwrap();
    let tmpl = env.get_template("sci6").unwrap();
    let rv = tmpl.render(context!());
    assert!(rv.is_ok());
}

#[test]
fn test_auto_escaping() {
    let mut env = Environment::new();
    env.add_template("index.html", "{{ var }}").unwrap();
    #[cfg(feature = "json")]
    {
        env.add_template("index.js", "{{ var }}").unwrap();
    }
    env.add_template("index.txt", "{{ var }}").unwrap();

    // html
    let tmpl = env.get_template("index.html").unwrap();
    let rv = tmpl.render(context!(var => "<script>")).unwrap();
    insta::assert_snapshot!(rv, @"&lt;script&gt;");

    // JSON
    #[cfg(feature = "json")]
    {
        use minijinja::value::Value;
        let tmpl = env.get_template("index.js").unwrap();
        let rv = tmpl.render(context!(var => "foo\"bar'baz")).unwrap();
        insta::assert_snapshot!(rv, @r###""foo\"bar'baz""###);
        let rv = tmpl
            .render(context!(var => [Value::from(true), Value::from("<foo>"), Value::from(())]))
            .unwrap();
        insta::assert_snapshot!(rv, @r###"[true,"<foo>",null]"###);
    }

    // Text
    let tmpl = env.get_template("index.txt").unwrap();
    let rv = tmpl.render(context!(var => "foo\"bar'baz")).unwrap();
    insta::assert_snapshot!(rv, @r###"foo"bar'baz"###);
}

#[test]
fn test_loop_changed() {
    let rv = minijinja::render!(
        r#"
        {%- for i in items -%}
          {% if loop.changed(i) %}{{ i }}{% endif %}
        {%- endfor -%}
        "#,
        items => vec![1, 1, 1, 2, 3, 4, 4, 5],
    );
    assert_eq!(rv, "12345");
}

// ideally this would work, but unfortunately the way serde flatten works makes it
// impossible for us to support with the internal optimizations in the value model.
// see https://github.com/mitsuhiko/minijinja/issues/222
#[derive(Debug, serde::Serialize)]
struct Bad {
    a: i32,
    #[serde(flatten)]
    more: Value,
}

#[test]
#[should_panic = "can only flatten structs and maps"]
fn test_flattening() {
    let ctx = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let env = Environment::new();
    env.render_str("{{ debug() }}", ctx).unwrap();
}

#[test]
fn test_flattening_sub_item_good() {
    let bad = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let ctx = context!(bad, good => "good");
    let env = Environment::new();

    // we are not touching a bad value, so we are good
    let rv = env.render_str("{{ good }}", ctx).unwrap();
    assert_eq!(rv, "good");
}

#[test]
#[should_panic = "can only flatten structs and maps"]
fn test_flattening_sub_item_bad_lookup() {
    let bad = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let ctx = context!(bad, good => "good");
    let env = Environment::new();

    // resolving an invalid value will fail
    env.render_str("{{ bad }}", ctx).unwrap();
}

#[test]
fn test_flattening_sub_item_bad_attr() {
    let bad = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let ctx = context!(good => context!(bad));
    let env = Environment::new();

    // resolving an invalid value will fail, even in an attribute lookup
    let err = env
        .render_str("{% if good.bad %}...{% endif %}", ctx)
        .unwrap_err();
    assert_eq!(err.kind(), ErrorKind::BadSerialization);
    assert_eq!(
        err.detail(),
        Some("can only flatten structs and maps (got an enum)")
    );
}

#[test]
fn test_flattening_sub_item_shielded_print() {
    let bad = Bad {
        a: 42,
        more: Value::from(BTreeMap::from([("b", 23)])),
    };

    let ctx = context!(good => context!(bad));
    let env = Environment::new();

    // this on the other hand is okay
    let value = env.render_str("{{ good }}", ctx).unwrap();
    assert_eq!(
        value,
        r#"{"bad": <invalid value: could not serialize to value: can only flatten structs and maps (got an enum)>}"#
    );
}

#[test]
#[cfg(feature = "custom_syntax")]
fn test_custom_syntax() {
    let mut env = Environment::new();
    env.set_syntax(
        minijinja::syntax::SyntaxConfig::builder()
            .block_delimiters("{", "}")
            .variable_delimiters("${", "}")
            .comment_delimiters("{*", "*}")
            .build()
            .unwrap(),
    );

    // this on the other hand is okay
    let value = env
        .render_str("{for x in range(3)}${x}{endfor}{* nothing *}", ())
        .unwrap();
    assert_eq!(value, r"012");
}

#[test]
fn test_undeclared_variables() {
    let mut env = Environment::new();
    env.add_template(
        "demo",
        "{% set x = foo %}{{ x }}{{ bar.baz }}{{ bar.blub }}",
    )
    .unwrap();
    let tmpl = env.get_template("demo").unwrap();
    let undeclared = tmpl.undeclared_variables(false);
    assert_eq!(
        undeclared,
        ["foo", "bar"].into_iter().map(|x| x.to_string()).collect()
    );
    let undeclared = tmpl.undeclared_variables(true);
    assert_eq!(
        undeclared,
        ["foo", "bar.baz", "bar.blub"]
            .into_iter()
            .map(|x| x.to_string())
            .collect()
    );
}

#[test]
fn test_undeclared_variables_bug() {
    // see https://github.com/mitsuhiko/minijinja/issues/484
    let mut env = Environment::new();
    env.add_template(
        "demo",
        r#"
        {% for item in content.products.items() %}
          {{ foo.bar }}
          {{ item.name }}  // this used to cause a loop
        {% endfor %}
    "#,
    )
    .unwrap();
    let tmpl = env.get_template("demo").unwrap();
    let undeclared = tmpl.undeclared_variables(true);
    assert_eq!(
        undeclared,
        ["foo.bar", "content.products.items"]
            .into_iter()
            .map(|x| x.to_string())
            .collect()
    );
}

#[test]
fn test_block_fragments() {
    let mut env = Environment::new();
    env.add_template(
        "demo",
        "I am outside the fragment{% block foo %}foo{% endblock %}So am I!",
    )
    .unwrap();
    let tmpl = env.get_template("demo").unwrap();

    let rv_a = tmpl.render(()).unwrap();
    let rv_b = tmpl.eval_to_state(()).unwrap().render_block("foo").unwrap();

    assert_eq!(rv_a, "I am outside the fragmentfooSo am I!");
    assert_eq!(rv_b, "foo");
}

#[test]
fn test_state() {
    let mut env = Environment::new();
    env.add_template(
        "foo.html",
        r#"
        {% set global = variable * 2 %}
        {% macro something() %}{{ global }}{% endmacro %}
        {% block baz %}[{{ global }}]{% endblock %}
    "#,
    )
    .unwrap();
    let template = env.get_template("foo.html").unwrap();
    let mut state = template
        .eval_to_state(context! {
            variable => 23
        })
        .unwrap();
    assert!(state.lookup("range").is_some());
    assert!(!state.exports().contains(&"range"));
    assert_eq!(state.lookup("global"), Some(Value::from(23 * 2)));
    assert_eq!(state.call_macro("something", &[]).unwrap(), "46");
    assert_eq!(state.render_block("baz").unwrap(), "[46]");
}

#[test]
#[allow(unused_mut)]
fn test_render_and_return_state() {
    let mut env = Environment::new();
    #[cfg(feature = "fuel")]
    {
        env.set_fuel(Some(100));
    }
    let tmpl = env
        .template_from_str("{% for x in range(3) %}Hello {{ name }}!\n{% endfor %}{% set x = 1 %}")
        .unwrap();
    let (rv, state) = tmpl
        .render_and_return_state(context! { name => "Foo" })
        .unwrap();
    assert_eq!(rv, "Hello Foo!\nHello Foo!\nHello Foo!\n");
    assert_eq!(state.lookup("x"), Some(Value::from(1)));

    #[cfg(feature = "fuel")]
    {
        assert_eq!(state.fuel_levels(), Some((26, 74)));
    }
}

#[test]
fn test_render_to_write_state() {
    let env = Environment::new();
    let tmpl = env
        .template_from_str("{% set foo = 42 %}{% macro bar() %}x{% endmacro %}root")
        .unwrap();
    let mut out = Vec::<u8>::new();
    let state = tmpl.render_to_write((), &mut out).unwrap();
    assert_eq!(String::from_utf8_lossy(&out), "root");
    assert_eq!(state.lookup("foo"), Some(Value::from(42)));
    assert_eq!(state.call_macro("bar", &[]).ok().as_deref(), Some("x"));
}

#[test]
fn test_functions() {
    assert_snapshot!(
        render!("{{ f() }}", f => Value::from_function(|| -> i32 { 42 })),
        @"42"
    );
    assert_snapshot!(
        render!("{{ f() }}", f => Value::from_function(|| -> Option<i32> { None })),
        @"none"
    );
    assert_snapshot!(
        render!("{{ f() }}", f => Value::from_function(|| -> Result<i32, Error> { Ok(23) })),
        @"23"
    );
}

#[test]
fn test_invalid_value_iteration() {
    let env = Environment::new();
    let mut out = Vec::<u8>::new();

    #[derive(Debug)]
    struct FailingIteration;

    impl Object for FailingIteration {
        fn repr(self: &Arc<Self>) -> ObjectRepr {
            ObjectRepr::Iterable
        }

        fn enumerate(self: &Arc<Self>) -> Enumerator {
            Enumerator::Iter(Box::new((0..4).map(Value::from).chain(Some(Value::from(
                Error::new(ErrorKind::InvalidOperation, "failed during iteration"),
            )))))
        }
    }

    let t = env
        .template_from_str("{% for item in iter %}[{{ item }}]{% endfor %}")
        .unwrap();
    let err = t
        .render_to_write(
            context! { iter => Value::from_object(FailingIteration) },
            &mut out,
        )
        .unwrap_err();
    let out = String::from_utf8_lossy(&out);
    assert_eq!(out, "[0][1][2][3]");
    assert_eq!(err.kind(), ErrorKind::InvalidOperation);
    assert_eq!(err.detail(), Some("failed during iteration"));
}
