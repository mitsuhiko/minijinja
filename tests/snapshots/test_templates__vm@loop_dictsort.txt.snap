---
source: tests/test_templates.rs
expression: "&rendered"
input_file: tests/inputs/loop_dictsort.txt

---

* 42: aha
* bar: 2
* baz: 100
* foo: 1

=====

Template {
    name: "loop_dictsort.txt",
    instructions: [
        00000 | LOOKUP (var "map")   [<unknown>:1],
        00001 | BUILD_LIST (0 items)   [<unknown>:1],
        00002 | APPLY_FILTER (name "dictsort")   [<unknown>:1],
        00003 | PUSH_LOOP (loop var: true)   [<unknown>:1],
        00004 | ITERATE (exit to 00010)   [<unknown>:1],
        00005 | UNPACK_LIST (2 items)   [<unknown>:1],
        00006 | STORE_LOCAL (var "key")   [<unknown>:1],
        00007 | STORE_LOCAL (var "value")   [<unknown>:1],
        00008 | EMIT_RAW (string "\n* ")   [<unknown>:1],
        00009 | LOOKUP (var "key")   [<unknown>:2],
        0000a | EMIT   [<unknown>:2],
        0000b | EMIT_RAW (string ": ")   [<unknown>:2],
        0000c | LOOKUP (var "value")   [<unknown>:2],
        0000d | EMIT   [<unknown>:2],
        0000e | EMIT_RAW (string "")   [<unknown>:2],
        0000f | JUMP (to 00004)   [<unknown>:2],
        00010 | POP_FRAME   [<unknown>:2],
        00011 | EMIT_RAW (string "\n")   [<unknown>:3],
    ],
    blocks: {},
    initial_auto_escape: None,
}
